### ISSUE TITLE
Inferred object type loses coverage after && operator

### ISSUE THREAD

**Issue by dmnd** [facebook/flow#2853]
I'm not sure what's going on here, but annotating the object makes it work properly, so maybe it's something to do with sealed vs unsealed objects. I'm still confused Flow only misses the error after the use of the && operator, though.

```js
const inferred = {
  x: '',
  y: '',
};

const a = inferred.y1; // error, as expected
const b = inferred.x && inferred.y1; // no error?

const declared = {
  x: ('': string),
  y: ('': string),
};

const c = declared.x && declared.y1; // error
const d = declared.y1; // error
```

See [Try Flow](https://flowtype.org/try/#0MYewdgzgLgBAlmAZgUwE6uQExgXhgbwCgYYAPALhgHIqAaYmAT0pvoF8BuQw0SWAQ1zwkaDJgB0jAIwcYAejkxRIVLRj8IS0gAdkwKFh7hoMAEZCEKdFnGkYAMnvCrYyTPmKwIJehUB+bl4TTD0AG34xISISChgAChpKaFQEAHMASnoSZnjEmGS0zMJOQOMBIRDgcNc7RxhK6ptpWQUfVBUjPjMKsIim91blVCA).

**Comment by samwgoldman** [comment]
Flow is being a bit clever here. In the inferred case, since we know that `inferred.x` is `''`, and thus definitely falsey, the `inferred.y1` in `inferred.x && inferred.y1` is unreachable and thus not an error.

Compare the following:

```
const o = {p:''};
o.p && o.q;
```

to

```
const o = {p:'non-empty'};
o.p && o.q;
```

In the latter we know that `o.p` is truthy, so `o.q` is reachable and we error.

Similarly, when `o` has type `{p:string}`, flow "forgets" that `o.p` has the literal value `''` and thus can't statically determine whether `o.p` is truthy or falsey and considers `o.q` reachable as well.

Long story short, this is expected and sound behavior.

**Comment by dmnd** [comment]
Ah, interesting. [Here's the original context](https://flowtype.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc65gA7TJKIgEziQA8YkzTo2ToYAOgDCuSM0HwA3uThw6MFALgBeOEpUreALgL4ANMv0BPY-jMWAvhQsALFEOpIACkQYAKAJS6FiowzsB04nRIMADK6gK+CkYEMFAArqGW+Pb+FCr2lCpEQuwBQfpwaBDManAo2nCh4ZHxSOKWAIykcAD0PTwc0KZ1jHxgSBhInMGV1bUARg1NEWoabbxwAGSbjWErre1dvf3MEAM4UAD8DuT2QA) in which I encountered it, which explains why I think this might still be a problem:

```js
import React from 'react';

class inferred extends React.Component {
  state = {
    x: '',
    y: '',
  };

  handlePress() {
    this.setState({x: 'truthy'});
  }

  render() {
    const a = this.state.y1; // error, as expected
    const b = this.state.x && this.state.y1; // no error?
  }
}
```

In the above context, I think I should an error in one of the following places:

 * `setState({x: 'truthy literal'})`
 * `this.state.x && this.state.y1`

**Comment by samwgoldman** [comment]
Ah, this is definitely a bug. We shouldn't assume that this.state.x is the literal string `''` here. Thanks for the report!

**Comment by gkz** [comment]
Errors now: https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0STMjAglAkAAIAEqsZRcmCUdJcgDkUWUooA3AAdKBylbGLkAZUsCy5AF4uQASPnsADyUEGAB5gPJkFz7KYoAheFzaBarRUEPSAHyy+WPXAYEgkLkVTHUUpchQLKBsP165RWADC6XZXx8psZKrV2FdXOAcq5lrTFtVmWwmsz2ZzXPNYtFvFLOYdlerUBz9I9pdqwLYbQACixfQAKACUJcbZYktVpViJBYWvbNFtFUlStVoovp-Y9TblpZY4ZMA6HZa5tnsXIwxdH4-shastAAjFKuQB6B8h6gcu0+kNGOYLNg1w-QY80DPMcSAnNMrHkLkADIoK5c9QMvBZrzvR9nygCAX2FSgAH5S3pOV6VyEAGhMEg4GgJIGgABisAAmAA2ABmKwqJAekgA
